File: settlement/room/food/cannibal/Cage.java
package settlement.room.food.cannibal;

import settlement.main.SETT;
import settlement.room.main.util.RoomBits;
import snake2d.util.bit.Bits;
import snake2d.util.datatypes.COORDINATE;
import snake2d.util.datatypes.Coo;

public class Cage{
	
	private final ROOM_CANNIBAL b;
	private CannibalInstance ins;
	private Coo coo = new Coo();
	

	private final int sNone = 0;
	private final int sReserved = 1;
	private final int sInside = 2;
	private final int sFetching = 3;
	
	private final RoomBits state = new RoomBits(coo, new Bits(0b0111)) {
		
		@Override
		protected void remove() {
			if (get() > sNone) {
				ins.prisoners --;
				b.prisoners --;
			}
			if (get() == sInside) {
				ins.reservable --;
			}
			
		};
		
		@Override
		protected void add() {
			if (get() > sNone) {
				ins.prisoners ++;
				b.prisoners ++;
			}
			if (get() == sInside) {
				ins.reservable ++;
			}
		};
		
	};

	Cage(ROOM_CANNIBAL print) {
		this.b = print;
	}
	
	Cage get(int tx, int ty) {
		
		ins = b.get(tx, ty);
		if (ins != null && SETT.ROOMS().fData.tile.is(tx, ty, ins.blueprintI().constructor.cc)) {
			coo.set(tx, ty);
			return this;
		}
		return null;
	}
	
	public boolean available() {
		return state.get() == sNone;
	}
	
	public void prisonerReserve() {
		state.set(ins, sReserved);
	}
	
	public void prisonerArrive() {
		state.set(ins, sInside);
	}
	
	public boolean prisonerOk() {
		return state.get() != sNone; 
	}
	
	public void prisonerCancel() {
		state.set(ins, sNone);
	}
	
	public boolean canGrab() {
		return state.get() == sInside;
	}
	
	public void grab() {
		state.set(ins, sFetching);
	}
	
	public void grabCancel() {
		if (state.get() == sFetching)
			state.set(ins, sInside);
	}
	
	public COORDINATE coo() {
		return coo;
	}
	
}

File: LegacySource/settlement/room/food/cannibal/Cage.cs
using System;
using snake2d.util.bit;
using snake2d.util.datatypes;

namespace settlement.room.food.cannibal
{
    public class Cage
    {
        private readonly ROOM_CANNIBAL b;
        private CannibalInstance ins;
        private Coo coo = new Coo();

        private readonly int sNone = 0;
        private readonly int sReserved = 1;
        private readonly int sInside = 2;
        private readonly int sFetching = 3;

        private readonly RoomBits state = new RoomBits(coo, new Bits(0b0111))
        {
            protected override void Remove()
            {
                if (Get() > sNone)
                {
                    ins.prisoners--;
                    b.prisoners--;
                }
                if (Get() == sInside)
                {
                    ins.reservable--;
                }
            }

            protected override void Add()
            {
                if (Get() > sNone)
                {
                    ins.prisoners++;
                    b.prisoners++;
                }
                if (Get() == sInside)
                {
                    ins.reservable++;
                }
            }
        };

        public Cage(ROOM_CANNIBAL print)
        {
            this.b = print;
        }

        public Cage Get(int tx, int ty)
        {
            ins = b.Get(tx, ty);
            if (ins != null && SETT.ROOMS().fData.tile.Is(tx, ty, ins.BlueprintI().constructor.cc))
            {
                coo.Set(tx, ty);
                return this;
            }
            return null;
        }

        public bool Available()
        {
            return state.Get() == sNone;
        }

        public void PrisonerReserve()
        {
            state.Set(ins, sReserved);
        }

        public void PrisonerArrive()
        {
            state.Set(ins, sInside);
        }

        public bool PrisonerOk()
        {
            return state.Get() != sNone;
        }

        public void PrisonerCancel()
        {
            state.Set(ins, sNone);
        }

        public bool CanGrab()
        {
            return state.Get() == sInside;
        }

        public void Grab()
        {
            state.Set(ins, sFetching);
        }

        public void GrabCancel()
        {
            if (state.Get() == sFetching)
                state.Set(ins, sInside);
        }

        public COORDINATE Coo()
        {
            return coo;
        }
    }
}